src/oop/special_class_method/hard/q4/MainTest.java

```diff
@@ -0,0 +1,238 @@
+                "Your " + string2FieldName + " does not have the correct type.");
+        assertTrue(testClass.hasModifier(string2FieldName, "private"),
+                "Your " + string2FieldName + " does not have the correct visibility modifier.");
+        assertTrue(testClass.hasField(stringFieldName),
+                "The attribute " +  stringFieldName + " could not be found in your class. Please make sure you have added it, it is spelled correctly.");
+        assertTrue(testClass.hasField(stringFieldName, String.class),
+                "Your " + stringFieldName + " does not have the correct type.");
+        assertTrue(testClass.hasModifier(stringFieldName, "private"),
+                "Your " + stringFieldName + " does not have the correct visibility modifier.");
+    }
+
+    private static Stream<Arguments> constructorInputProvider() {
+        return Stream.of(
+                Arguments.of("Test String", "Long"),
+                Arguments.of("MyString", "Round"),
+                Arguments.of("First", "Second"),
+                Arguments.of("Third", "Fourth")
+
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    public void detergentClassHasCorrectConstructor(String value1, String value2) throws Throwable {
+        Object[][] arguments = {
+                {value1, String.class},
+                {value2, String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        String incorrectString1ValueMessage =
+                "Your " + className + " constructor does not initialize the " + string2FieldName + " attribute to the correct value.";
+        String incorrectString2ValueMessage =
+                "Your " + className + " constructor does not initialize the " + stringFieldName + " attribute to the correct value.";
+        _assertEquals(value1, testClass.getFieldValue(classInstance, string2FieldName),
+                incorrectString1ValueMessage);
+        _assertEquals(value2, testClass.getFieldValue(classInstance, stringFieldName),
+                incorrectString2ValueMessage);
+    }
+
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    public void detergentClassHasCorrectSetNameMethod(String initialValue, String setValue) throws Throwable {
+        Object[][] arguments = {
+                {initialValue, String.class},
+                {"Initial String", String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] setSizeArguments = {
+                {setValue, String.class}
+        };
+        String[] setMethodModifiers = {"public"};
+        String incorrectSetMethodMessage = String.join(" ",
+                "Your", setString2MethodName, "method does not correctly set the value of the", string2FieldName, "attribute.");
+        Object setMethodOutput = testClass.callMethod(setString2MethodName, setSizeArguments, setMethodModifiers, classInstance);
```

Comment:
use .hasMethod first with void.type instead of assertNull here and everywhere else